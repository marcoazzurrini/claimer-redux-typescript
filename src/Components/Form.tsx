/** @jsx jsx */
import { jsx } from "@emotion/core";
import { useState } from "react";
import { formStyles } from "../Styles/form";
import useFormValidation from "../Hooks/useFormValidation";
import PropTypes from "prop-types";
import Button from "./Button";
import { BtnMedium, messageStyles } from "../Styles/buttons";

interface Iquestion {
  id: string;
  question: string;
  tag: string;
  answers: [];
  type: string;
}
interface props {
  handleFormSubmit: Function;
  render: any;
  currentQuestion: Iquestion;
}

export default function Form({
  handleFormSubmit,
  render,
  currentQuestion
}: props) {
  const [inputVal, setInputVal] = useState<string>("");
  const [error, setError] = useState(false);

  const isValidated = useFormValidation(inputVal, currentQuestion.type);

  const handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();

    if (isValidated && inputVal !== "") {
      handleFormSubmit(inputVal);
      setInputVal("");
    } else {
      setError(true);
      setTimeout(() => setError(false), 2500);
    }
  };

  const handleChange = (inputValue: string) => setInputVal(inputValue);

  return (
    <form css={formStyles} onSubmit={handleSubmit}>
      {render({ inputVal, handleChange, currentQuestion })}
      {error ? (
        <Button style={messageStyles}>Please enter a valid input</Button>
      ) : null}
      <Button style={BtnMedium} type="submit">
        Submit
      </Button>
    </form>
  );
}

Form.propTypes = {
  render: PropTypes.func.isRequired,
  handleFormSubmit: PropTypes.func.isRequired,
  currentQuestion: PropTypes.object.isRequired
};
