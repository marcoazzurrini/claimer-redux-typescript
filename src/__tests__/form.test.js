import React from "react";
import Form from "../Components/Form";
import { shallow, mount } from "enzyme";
import QuizQuestions from "../Api/questions";
const currentQuestion = QuizQuestions[0];

const MockElement = props => <input />;
describe("testing Form component", () => {
  it("should render without errors", () => {
    shallow(
      <Form
        currentQuestion={currentQuestion}
        handleSubmit={() => null}
        render={stuff => <MockElement stuff={stuff} />}
      />
    );
  });

  it("should match snapshot", () => {
    const wrapper = shallow(
      <Form
        currentQuestion={currentQuestion}
        handleSubmit={() => null}
        render={stuff => <MockElement stuff={stuff} />}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("should render one form", () => {
    const wrapper = mount(
      <Form
        currentQuestion={currentQuestion}
        handleSubmit={() => null}
        render={stuff => <MockElement stuff={stuff} />}
      />
    );
    expect(wrapper.find("form")).toHaveLength(1);
  });

  it("should render one button", () => {
    const wrapper = mount(
      <Form
        currentQuestion={currentQuestion}
        handleSubmit={() => null}
        render={stuff => <MockElement stuff={stuff} />}
      />
    );
    expect(wrapper.find("button")).toHaveLength(1);
  });

  it("should render one input if passed throgh the render prop", () => {
    const wrapper = mount(
      <Form
        currentQuestion={currentQuestion}
        handleSubmit={() => null}
        render={stuff => <MockElement stuff={stuff} />}
      />
    );
    expect(wrapper.find("input")).toHaveLength(1);
  });
});
