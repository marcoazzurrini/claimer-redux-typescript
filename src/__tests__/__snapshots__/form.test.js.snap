// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing Form component should match snapshot 1`] = `ShallowWrapper {}`;
